File: apps/android/app/src/main/java/com/studypop/android/data/StudyPopApi.kt
package com.studypop.android.data

import com.studypop.android.BuildConfig
import com.studypop.android.model.ChatMessage
import com.studypop.android.model.StudyKit
import com.studypop.android.model.StudyPopState
import com.studypop.android.model.StudyCompanion
import com.studypop.android.model.StudySection
import org.json.JSONArray
import org.json.JSONObject

class StudyPopApi(private val auth: FirebaseAuthService) {
    private val baseUrl = BuildConfig.STUDYPOP_API_BASE_URL.trimEnd('/')
    private var stateVersion = 0

    suspend fun bootstrap() = run {
        val config = request("/api/v1/config")
        val firebase = config.optJSONObject("firebase")
        val publicConfig = firebase?.optJSONObject("config")
        if (firebase?.optBoolean("configured") != true || publicConfig == null) {
            throw StudyPopException(
                "StudyPop accounts are temporarily unavailable.",
                statusCode = 503,
                code = "FIREBASE_NOT_CONFIGURED",
            )
        }
        auth.configure(publicConfig.getString("apiKey"))
        auth.restoreUser()
    }

    suspend fun loadState(): StudyPopState? {
        val response = request("/api/v1/state", authenticated = true)
        stateVersion = response.optInt("version")
        return response.optJSONObject("state")?.let(StudyPopState::fromJson)
    }

    suspend fun saveState(state: StudyPopState) {
        val response = request(
            path = "/api/v1/state",
            method = "PUT",
            body = JSONObject()
                .put("state", state.toJson())
                .put("version", stateVersion),
            authenticated = true,
        )
        stateVersion = response.optInt("version", stateVersion + 1)
    }

    suspend fun deleteState() {
        request("/api/v1/state", method = "DELETE", authenticated = true)
        stateVersion = 0
    }

    suspend fun answer(
        section: StudySection,
        question: String,
        images: List<String>,
        history: List<ChatMessage>,
        companion: StudyCompanion,
        studyKit: StudyKit?,
    ): String {
        val body = JSONObject()
            .put("section", section.id)
            .put("question", question)
            .put("images", JSONArray(images))
            .put(
                "history",
                JSONArray().apply {
                    history.takeLast(8).forEach {
                        put(JSONObject().put("role", it.role).put("text", it.text))
                    }
                },
            )
            .put("companion", companion.displayName)
            .put("studyContext", studyKit?.toJson() ?: JSONObject.NULL)
        return request(
            path = "/api/v1/ai/answer",
            method = "POST",
            body = body,
            authenticated = auth.currentUser != null,
        ).getString("answer")
    }

    suspend fun createStudyKit(note: String, images: List<String>): StudyKit {
        val body = JSONObject()
            .put("note", note)
            .put("images", JSONArray(images))
        val response = request(
            path = "/api/v1/ai/study-kit",
            method = "POST",
            body = body,
            authenticated = auth.currentUser != null,
        )
        return StudyKit.fromJson(response.getJSONObject("kit"))
    }

    suspend fun transcribe(audioDataUrl: String, mimeType: String): String {
        val response = request(
            path = "/api/v1/ai/transcribe",
            method = "POST",
            body = JSONObject()
                .put("audio", audioDataUrl)
                .put("mimeType", mimeType),
            authenticated = auth.currentUser != null,
        )
        return response.getString("text")
    }

    fun resetVersion() {
        stateVersion = 0
    }

    private suspend fun request(
        path: String,
        method: String = "GET",
        body: JSONObject? = null,
        authenticated: Boolean = false,
    ): JSONObject {
        val execute: suspend (Boolean) -> JSONObject = { forceRefresh ->
            val token = if (authenticated) auth.idToken(forceRefresh) else null
            HttpClient.request(
                url = "$baseUrl$path",
                method = method,
                body = body?.toString(),
                bearerToken = token,
            )
        }
        return try {
            execute(false)
        } catch (error: StudyPopException) {
            if (authenticated && error.statusCode == 401) execute(true) else throw error
        }
    }
}

File: apps/ios/StudyPop/Services/StudyPopAPI.swift
import Foundation

actor StudyPopAPI {
    private struct ConfigResponse: Decodable {
        struct FirebaseBlock: Decodable {
            struct PublicConfig: Decodable {
                let apiKey: String
            }
            let configured: Bool
            let config: PublicConfig?
        }
        let firebase: FirebaseBlock
    }

    private struct StateResponse: Decodable {
        let state: StudyPopState?
        let version: Int
        let updatedAt: String?
    }

    private struct SaveResponse: Decodable {
        let saved: Bool
        let version: Int
        let updatedAt: String?
    }

    private struct AnswerResponse: Decodable {
        let answer: String
    }

    private struct StudyKitResponse: Decodable {
        let kit: StudyKit
    }

    private struct TranscriptionResponse: Decodable {
        let text: String
    }

    private let baseURL: URL
    private let auth: FirebaseAuthService
    private var stateVersion = 0

    init(auth: FirebaseAuthService) {
        let configured = Bundle.main.object(
            forInfoDictionaryKey: "STUDYPOP_API_BASE_URL"
        ) as? String
        baseURL = URL(string: configured ?? "https://studypop-flame.vercel.app")!
        self.auth = auth
    }

    func bootstrap() async throws -> AuthUser? {
        let config: ConfigResponse = try await request(path: "/api/v1/config")
        guard config.firebase.configured, let firebase = config.firebase.config else {
            throw APIError(
                message: "StudyPop accounts are temporarily unavailable.",
                statusCode: 503,
                code: "FIREBASE_NOT_CONFIGURED"
            )
        }
        await auth.configure(apiKey: firebase.apiKey)
        return await auth.currentUser()
    }

    func loadState() async throws -> StudyPopState? {
        let response: StateResponse = try await request(
            path: "/api/v1/state",
            authenticated: true
        )
        stateVersion = response.version
        return response.state
    }

    func saveState(_ state: StudyPopState) async throws {
        struct Body: Encodable {
            let state: StudyPopState
            let version: Int
        }
        let response: SaveResponse = try await request(
            path: "/api/v1/state",
            method: "PUT",
            body: Body(state: state, version: stateVersion),
            authenticated: true
        )
        stateVersion = response.version
    }

    func answer(
        section: StudySection,
        question: String,
        images: [String],
        history: [ChatMessage],
        companion: Companion,
        studyKit: StudyKit?
    ) async throws -> String {
        struct Body: Encodable {
            struct HistoryMessage: Encodable {
                let role: String
                let text: String
            }
            let section: String
            let question: String
            let images: [String]
            let history: [HistoryMessage]
            let companion: String
            let studyContext: StudyKit?
        }
        let body = Body(
            section: section.rawValue,
            question: question,
            images: images,
            history: history.suffix(8).map {
                .init(role: $0.role.rawValue, text: $0.text)
            },
            companion: companion.name,
            studyContext: studyKit
        )
        let response: AnswerResponse = try await request(
            path: "/api/v1/ai/answer",
            method: "POST",
            body: body,
            authenticated: await auth.currentUser() != nil
        )
        return response.answer
    }

    func createStudyKit(note: String, images: [String]) async throws -> StudyKit {
        struct Body: Encodable {
            let note: String
            let images: [String]
        }
        let response: StudyKitResponse = try await request(
            path: "/api/v1/ai/study-kit",
            method: "POST",
            body: Body(note: note, images: images),
            authenticated: await auth.currentUser() != nil
        )
        return response.kit
    }

    func transcribe(audioDataURL: String, mimeType: String) async throws -> String {
        struct Body: Encodable {
            let audio: String
            let mimeType: String
        }
        let response: TranscriptionResponse = try await request(
            path: "/api/v1/ai/transcribe",
            method: "POST",
            body: Body(audio: audioDataURL, mimeType: mimeType),
            authenticated: await auth.currentUser() != nil
        )
        return response.text
    }

    func resetVersion() {
        stateVersion = 0
    }

    private func request<Response: Decodable>(
        path: String,
        method: String = "GET",
        authenticated: Bool = false
    ) async throws -> Response {
        try await request(
            path: path,
            method: method,
            bodyData: nil,
            authenticated: authenticated
        )
    }

    private func request<Response: Decodable, Body: Encodable>(
        path: String,
        method: String,
        body: Body,
        authenticated: Bool
    ) async throws -> Response {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        return try await request(
            path: path,
            method: method,
            bodyData: encoder.encode(body),
            authenticated: authenticated
        )
    }

    private func request<Response: Decodable>(
        path: String,
        method: String,
        bodyData: Data?,
        authenticated: Bool
    ) async throws -> Response {
        var request = URLRequest(url: baseURL.appending(path: path))
        request.httpMethod = method
        if let bodyData {
            request.httpBody = bodyData
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        if authenticated {
            request.setValue(
                "Bearer \(try await auth.idToken())",
                forHTTPHeaderField: "Authorization"
            )
        }

        let (data, response) = try await URLSession.shared.data(for: request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 500
        if status == 401, authenticated {
            request.setValue(
                "Bearer \(try await auth.idToken(forceRefresh: true))",
                forHTTPHeaderField: "Authorization"
            )
            let (retryData, retryResponse) = try await URLSession.shared.data(for: request)
            return try decode(retryData, response: retryResponse)
        }
        return try decode(data, response: response)
    }

    private func decode<Response: Decodable>(
        _ data: Data,
        response: URLResponse
    ) throws -> Response {
        let status = (response as? HTTPURLResponse)?.statusCode ?? 500
        guard (200..<300).contains(status) else {
            let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
            let rawError = object?["error"]
            let payload = rawError as? [String: Any]
            let message = payload?["message"] as? String
                ?? rawError as? String
                ?? "StudyPop could not complete that request."
            throw APIError(
                message: message,
                statusCode: status,
                code: payload?["code"] as? String ?? "API_REQUEST_FAILED"
            )
        }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode(Response.self, from: data)
    }
}
